### ISSUE TITLE
Downloads Break More Easily On 2.5.0

### ISSUE THREAD

**Issue by BeranabusBarnes** [ModOrganizer2/modorganizer#1989]
## The problem:
Queuing up multiple downloads via mod manager download button on Nexus breaks downloads in MO2. Download slow down or freeze completely. Pausing and resuming not help. Only fix is delete partial download and redownload completely.

In previous versions, downloads could mess up and freeze entire application. Restart and resume download work though.

## To Reproduce:

1. Go to multiple mod pages.
2. Hit mod manager download on all of them.

## Environment:

* Mod Organizer Version that exhibits the issue: 2.5.0
* Last Mod Organizer Version that did not exhibit the issue (if applicable): 2.4.4 (but had its own problems)
* Desktop OS/version used to run Mod Organizer: W10

## Details:
Have always had a lot of issues with download function in MO2.
Downloads kept on different hard disk drive.
Have a ton of downloads.

## Link to Mod Organizer logs:
Right click on the log window at the bottom of MO2, set log level to debug, restart Mo2 and reproduce the problem.
To access the log files you can press the Open folder icon over the modlist and select "Open MO2 Logs folder".

### USVFS:
[Link](https://gist.github.com/BeranabusBarnes/8bd20ef98371bad939baa867a12d92ec)

### MO Interface:
[Link](https://gist.github.com/BeranabusBarnes/a96a6286e7f16aafce516e6d4fe3d210)

**Comment by BeranabusBarnes** [comment]
Workaround found for all download issues: download manually into downloads folder and query after. Having so many problems with MO2 downloads lately.

**Comment by sougeki** [comment]
+1 to that
I made a whole build with v2.4.4, haven't had a single download break. Switched to v2.5.0, and the first consecutive (within a 2~3s timeframe) download I tried, broke.
I'll revert back to 2.4.4 for now, I hope it gets fixed in the next version

**Comment by krypto5863** [comment]
Confirmed, also having this issue. Updates constantly stop.

**Comment by hikingmetalpunk** [comment]
I think I'm seeing this as well; it works better but not perfectly if I only attempt to download one at a time.

**Comment by WhisperySoap** [comment]
Just started for me yesterday. 2 weeks ago, I downloaded over 200 mods with no issues. Yesterday downloading any mods over 200kb or more then 1 at a time is nearly guaranteed to randomly freeze, finish corrupted, or require 3-7 retries.

**Comment by Sixdd6** [comment]
I'm also seeing this behavior. It seems to be more common when there are more than a few downloads in the list. I just started making a modlist today and after the first five or so went smoothly the next ten all stalled when adding them through the nexus "download with manager" link, meaning they would add the mod to the list but the download never starts unless you right-click and pause then un-pause.

**Comment by xSwagboy69x** [comment]
Also experiencing this issue, more so since 2.5.0 (but it definitely happened on 2.4.4 as well).

Found an old issue thread from both 2018 and last year also mentioning it. Seems like it's a longstanding problem with MO2's download manager itself that only really starts to occur regularly when the user has a significant number of mods displayed in the downloads tab.

**Comment by Babkatt** [comment]
Having this same problem. Seems to worsen when attempting to download multiple files at once. Pausing/resuming a download sometimes causes MO2 to crash.

**Comment by hikingmetalpunk** [comment]
I haven't worked with modding in a while, frustrating to come back to it and find this is still an issue when some releases have happened in the meantime

**Comment by hikingmetalpunk** [comment]
> > I haven't worked with modding in a while, frustrating to come back to it and find this is still an issue when some releases have happened in the meantime
> 
> There isn't really a solution to this – starting a download sets off a chain reaction of:
> 
> Requesting the file from CDN, Querying mod data from API, Creating and Drawing new entry in download tab, Writing file to disk, Checking the API for available updates for the file, and then finally performing a full refresh and re-drawing/re-rendering of the download tab – and this final step gets slower and slower the more downloads you have in this tab, because that means more and more entries need to be redrawn. This last step involving the refresh and redraw within MO2 is t he crux of the problem, and a solution probably would require an entire re-write of Mod Organizer itself (speculating, of course).
> 
> Combine all of this with potentially slow hardware from the user, such as using a traditional HDD to store downloaded mods, and yeah it can get rough. I've personally made a habit or taking time between downloads, and not spamming downloads willy-nilly.

I may be misinterpreting the report then (or several mechanisms cause similar behavior) because I’m seeing downloads freeze irrespective of how many completed are in the tab or whether multiple are going at once.

**Comment by xSwagboy69x** [comment]
> I may be misinterpreting the report then (or several mechanisms cause similar behavior) because I’m seeing downloads freeze irrespective of how many completed are in the tab or whether multiple are going at once.

Yeah I understand, I was just pointing out all stuff that can be a cause for this, and how it can manifest in different ways. This problem has been my bane since I started using MO2 years ago and I've put more time into trying to learn this problem than I'd like to admit.

As for your specific issue, let me try to get some clarification from you to make sure I understand exactly what's happening on your end.

If I understand you correctly, you are hitting the download button one after the other on a mod page that has several (or more than several) files available – for example, a gameplay mod with tons of pre-made combability patches made by the author.

In normal working conditions, clicking the "download" button will then show a little (_non-Windows system_) popup at the _**center-bottom**_ of your screen, above the taskbar. If you see this, the download process has started normally and should end up with a successful file transfer.

**_However_**, if you then click to download _another_ subsequent mod in too quick a time frame, problems can start to happen. The file request probably will go through, but you will notice that the aforementioned MO2 popup will now appear _center-screen_, instead of _center-buttom_. This is (at least in my experience), a confirmation that something wrong went with the file request, and you can confirm this by switching to the MO2 app where you will see one (or more) _native windows system popups_ describing various errors. This leads to download entries without metadata, or files freezing as it tries to query metadata from the API. Sometimes even if the download does finish sucessfully, the file will freeze or hang during the subsequent API query that checks for updated version numbers, which can then freeze _other, subsequent_ downloads – causing a snowball effect.

And this is just the issues caused by network side – as I started in the previous reply, the whole refresh tab and re-draw entries mechanism can/will slow things down on the system side even further, drastically slowing down the entire process of trying to download mods from Nexus. I consider them to be different parts of the same problem, since they almost always appear together.

Some workaround that could help:

- Wait a reasonable time before clicking to download the next mod.
- Set your download folder location on an SSD. It makes a huge difference with certain parts of the download process, like the simultaneous download of .meta files and anything related to MO2's virtual file system.

Let me know if any of these insights are applicable to your situation.

Hope this helps!

**Comment by hikingmetalpunk** [comment]
> > I may be misinterpreting the report then (or several mechanisms cause similar behavior) because I’m seeing downloads freeze irrespective of how many completed are in the tab or whether multiple are going at once.
> 
> Yeah I understand, I was just pointing out all stuff that can be a cause for this, and how it can manifest in different ways. This problem has been my bane since I started using MO2 years ago and I've put more time into trying to learn this problem than I'd like to admit.
> 
> As for your specific issue, let me try to get some clarification from you to make sure I understand exactly what's happening on your end.
> 
> If I understand you correctly, you are hitting the download button one after the other on a mod page that has several (or more than several) files available – for example, a gameplay mod with tons of pre-made combability patches made by the author.
> 
> In normal working conditions, clicking the "download" button will then show a little (_non-Windows system_) popup at the _**center-bottom**_ of your screen, above the taskbar. If you see this, the download process has started normally and should end up with a successful file transfer.
> 
> **_However_**, if you then click to download _another_ subsequent mod in too quick a time frame, problems can start to happen. The file request probably will go through, but you will notice that the aforementioned MO2 popup will now appear _center-screen_, instead of _center-buttom_. This is (at least in my experience), a confirmation that something wrong went with the file request, and you can confirm this by switching to the MO2 app where you will see one (or more) _native windows system popups_ describing various errors. This leads to download entries without metadata, or files freezing as it tries to query metadata from the API. Sometimes even if the download does finish sucessfully, the file will freeze or hang during the subsequent API query that checks for updated version numbers, which can then freeze _other, subsequent_ downloads – causing a snowball effect.
> 
> And this is just the issues caused by network side – as I started in the previous reply, the whole refresh tab and re-draw entries mechanism can/will slow things down on the system side even further, drastically slowing down the entire process of trying to download mods from Nexus. I consider them to be different parts of the same problem, since they almost always appear together.
> 
> Some workaround that could help:
> 
>     * Wait a reasonable time before clicking to download the next mod.
> 
>     * Set your download folder location on an SSD. It makes a huge difference with certain parts of the download process, like the simultaneous download of .meta files and anything related to MO2's virtual file system.
> 
> 
> Let me know if any of these insights are applicable to your situation.
> 
> Hope this helps!

Appreciate you taking the time, here's my best attempt at describing what I'm seeing.

I go through the mod manager download steps on Nexus; in mod manager one of two things happens:
* The mod appears in the download panel on the right side of the screen, the progress bar ticks upwards, the mod eventually finishes downloading
* The mod appears in the download panel, but no download progress happens and the total size remains at 0 bytes. Closing mod manager and reopening it will change the mod's status to Paused at which point I can resume it and it downloads like normal.

Some things I've noticed:
* Happens regularly downloading single mods
* Happens consistently if I try to download more than one mod at a time
* Does not seem to hinge on how many mods are in the download panel
* Does not seem to happen in clusters, after certain amount of time using MO, or anything like that
* Does not seem to happen with very large (e.g., dozens of megs or higher) mods
* Once one mod download has frozen in this way all subsequent attempts will as well until MO restarts
* No file corruption observed; it seems that the download probably doesn't even start until after MO is restarted
